### PR TITLE
Switch to C99 without the features that became optional in C11

### PR DESCRIPTION
**Pull request by kroeckx**

### CODE REVIEW

File: policies/coding-style.md [hunk 1]
```diff
@@ -586,10 +586,8 @@ it portable enough to meet our multi-platform support goals.
 
 ## Chapter 14:  Portability
 
-To maximise portability the version of C defined in ISO/IEC 9899:1990
-should be used. This is more commonly referred to as C90. ISO/IEC 9899:1999
-(also known as C99) is not supported on some platforms that OpenSSL is
-used on and therefore should be avoided.
+The version of C defined in ISO/IEC 9899:1999 (C99) should be used.
+Features that became optional in ISO/IEC 9899:2011 (C11) should not be used.
```
Comment (t8m) [inline policies/coding-style.md:590]: Should this change apply to 3.1 and newer releases only?

**Comment by levitte** [comment]
How do we handle old compilers, on platforms where there isn't any choice?  Will it be sufficient to add compatibility functions and macros, or are we thinking of supporting syntactic things that just aren't possible on those (for example, field specifiers when setting the value of a union)?

File: policies/coding-style.md (hunk 1 above)
Comment (kroeckx) [reply policies/coding-style.md:590]: That was at least the intention. How do we handle such changes in other policies? just add some text saying it only applies to 3.1 and later?

**Comment by kroeckx** [comment]
We already have compatibility for such platforms. We have plenty of code that's there just because we want to use C99 and C11 features that are not available on all platforms. But as far as I know, we don't have a platform in our platform lists that doesn't support C99, which includes things like Sun C and djgpp. This change in policy will allow us to remove such compatibility.

**Comment by t-j-h** [comment]
I don't believe that there are any C99 features that are sufficiently important to move forward to that language version.
OpenSSL is used on a large range of platforms that go well beyond the official list. We would also have to continue with C89 level support for all current releases.

**Comment by hlandau** [comment]
What are the features in C99 we actually want?

Taking a list of features from Wikipedia:
>inline functions

We already have our own macro for this.

>intermingled declarations and code

Would be nice, but major stylistic change, so probably undesirable.

>several new [data types](https://en.wikipedia.org/wiki/Data_type), including long long int, optional extended integer types, an explicit [boolean data type](https://en.wikipedia.org/wiki/Boolean_datatype), and a complex type to represent [complex numbers](https://en.wikipedia.org/wiki/Complex_number):

Bool would be nice and easily polyfilled on platforms which don't have it. Complex probably has poor support and I can't see us needing it.  long long we already use.

>[flexible array members](https://en.wikipedia.org/wiki/Flexible_array_member)

This doesn't really buy us anything and probably has poor support.

>support for one-line [comments](https://en.wikipedia.org/wiki/Comment_(programming)) beginning with //, as in [BCPL](https://en.wikipedia.org/wiki/BCPL), [C++](https://en.wikipedia.org/wiki/C%2B%2B) and [Java](https://en.wikipedia.org/wiki/Java_(programming_language))

Would be nice, but major stylistic change, so probably undesirable.

>new library functions, such as snprintf

We use our own implementation of printf in BIO, so this doesn't really matter to us.

>designated initializers (for example, initializing a structure by field names: struct point p = { .x = 1, .y = 2 };)[[5]](https://en.wikipedia.org/wiki/C99#cite_note-5)
>compound literals (for instance, it is possible to construct structures in function calls: function((struct x) {1, 2}))[[6]](https://en.wikipedia.org/wiki/C99#cite_note-6)

Would be nice but not a game-changer. I think we would need to survey support on our target platforms (VMS, NonStop, ...)

>support for [variadic macros](https://en.wikipedia.org/wiki/Variadic_macro) (macros with a variable number of arguments)

Probably the most useful thing in C99 for us. This only affects the preprocessor.

>[restrict](https://en.wikipedia.org/wiki/Restrict)

Desirable. We can use a macro for this so we can compile it out on platforms that don't support it. We can adopt this whether or not we adopt C99 in general.

I think aside from creating an `ossl_restrict` macro the only things useful to us are `_Bool` and variadic macros. We could also  create an `ossl_static_assert` macro for platforms which support C11 static assert. Of these the only things which can't be polyfilled/no-opped on platforms which don't support them are variadic macros. So the question is whether all platforms can support them.

**Comment by beldmit** [comment]
Personally I like boolean, declaring cycle variables inside the cycle itself and //-style comments. `long long` looks like a bad idea to me, I prefer explicitly sized types.

File: policies/coding-style.md (hunk 1 above)
Comment (t8m) [reply policies/coding-style.md:590]: > just add some text saying it only applies to 3.1 and later?

Yes, I think so.

**Comment by t8m** [comment]
> `long long` looks like a bad idea to me, I prefer explicitly sized types.

We already need and use it on 32 bit platforms for 64 bit numbers. (And yeah mostly as a typedef to int64_t)

**Comment by levitte** [comment]
> > designated initializers (for example, initializing a structure by field names: struct point p = { .x = 1, .y = 2 };)[[5]](https://en.wikipedia.org/wiki/C99#cite_note-5)
> > compound literals (for instance, it is possible to construct structures in function calls: function((struct x) {1, 2}))[[6]](https://en.wikipedia.org/wiki/C99#cite_note-6)
> 
> Would be nice but not a game-changer. I think we would need to survey support on our target platforms (VMS, NonStop, ...)
> 
> > support for [variadic macros](https://en.wikipedia.org/wiki/Variadic_macro) (macros with a variable number of arguments)
> 
> Probably the most useful thing in C99 for us. This only affects the preprocessor.

These two would be useful, but seriously precludes older compilers / pre-processors, as these are syntax features that can't be replicated with smart functions or macros...
That being said, *current* VMS (v8.4) with a *current* C compiler (v7.4) [seems to fully support C99](https://docs.vmssoftware.com/vsi-c-language-reference-manual/)

**Comment by kroeckx** [comment]
My understanding of the current situation is that we currently use those unconditionally:
- long long int
- flexible array members (not sure if we still do)
- snprintf

We use those conditionally:
- inttypes (like uint64_t)
- inline functions

Things that look useful that we currently don't use:
- intermingled declarations and code
- booleans
- designated initializers
- variadic macros
- restrict
- Unicode

Things that some people might want to use:
- one-line comments using //.  The coding style doesn't allow it.

**Comment by hlandau** [comment]
I doubt Unicode is of use to us.

**Comment by h-vetinari** [comment]
> > several new [data types](https://en.wikipedia.org/wiki/Data_type), including long long int, optional extended integer types, an explicit [boolean data type](https://en.wikipedia.org/wiki/Boolean_datatype), and a complex type to represent [complex numbers](https://en.wikipedia.org/wiki/Complex_number):
> 
> Bool would be nice and easily polyfilled on platforms which don't have it. Complex probably has poor support and I can't see us needing it. long long we already use.

There's an issue for that (and some preliminary discussion on C99) already in https://github.com/openssl/openssl/issues/17494.

Also, nonstop would be fine with C99 based on [this](https://github.com/openssl/technical-policies/issues/33#issuecomment-1069580781) comment.

Sidenote: this PR should probably have "Closes #33" in the OP.

**Comment by kroeckx** [comment]
Unicode support is something that can be useful in the apps, but to properly use it we probably also need something like iconv().

**Comment by kroeckx** [comment]
An alternative could be that we list which C99 features that we require.

**Comment by t8m** [comment]
> An alternative could be that we list which C99 features that we require.

**OTC: We think that this is the only way forward. Approving C99 as a whole does not work for supporting older platforms that we still want to support.**

**Comment by t8m** [comment]
@kroeckx ping

**Comment by t8m** [comment]
**OTC: vote has not been called. Closing.**

**Pull request closed without merging**